pilot/pkg/model/sidecar.go

```diff
@@ -217,7 +217,53 @@ func ConvertToSidecarScope(ps *PushContext, sidecarConfig *Config, configNamespa
 	// Assign namespace dependencies
 	out.namespaceDependencies = make(map[string]struct{})
 	for _, listener := range out.EgressListeners {
-		for _, s := range listener.services {
+		// Explicitly callable services are potential destinations, copy those first
+		serviceDestinations := append([]*Service(nil), listener.services...)
+
+		addDestination := func(d *networking.Destination) {
+			if d == nil {
+				return
+			}
+			// Default to this hostname in our config namespace
+			s, ok := ps.ServiceByHostnameAndNamespace[host.Name(d.Host)][configNamespace]
+			if !ok {
+				// Otherwise pick randomly
+				for _, v := range ps.ServiceByHostnameAndNamespace[host.Name(d.Host)] {
+					s, ok = v, true
+				}
+				if !ok {
+					return
+				}
+			}
+
+			serviceDestinations = append(serviceDestinations, s)
+		}
+
+		// Infer more possible destinations from virtual services
+		for _, vs := range listener.virtualServices {
+			v := vs.Spec.(*networking.VirtualService)
```

Comment:
#20124 has this exact same logic, may be worth breaking out to a helper. Can be a follow up pr though

Reply:
+1